ENH Adds get_feature_names_out to neighbors module

Micky774 opened the pull request:
#### Reference Issues/PRs
Addresses #21308

#### What does this implement/fix? Explain your changes.
Implements the `get_feature_names_out` method to the neighbors module

#### Any other comments?
This is a work in progress

Micky774 committed:
Added `get_feature_names_out` support to neighbors module

Initial implementation of `get_feature_names_out`

Micky774 committed:
Brought `sklearn/neighbors/_base.py` into style compliance

Micky774 committed:
Added `_n_features_out` attribute to NeighborsBase

ogrisel commented:
For `KNeighborsTransformer`, `n_neighbors` is fixed a priori. That should answer your comment above.

thomasjpfan reviewed (commented):
Thanks for the PR!

Micky774 commented on sklearn/neighbors/_base.py:
```diff
@@ -322,7 +322,9 @@ def _radius_neighbors_from_graph(graph, radius, return_distance):
 
-class NeighborsBase(MultiOutputMixin, BaseEstimator, metaclass=ABCMeta):
+class NeighborsBase(
+    _ClassNamePrefixFeaturesOutMixin, MultiOutputMixin, BaseEstimator, metaclass=ABCMeta
```
There's also `RadiusNeighborsTransformer`. I'll move the mixin to the more specific transformer classes, thanks!

Micky774 committed:
Improved implementation of `get_feature_names_out`

- Moved `_ClassNamePrefixFeaturesOutMixin` to specific transformers
- Added `_ClassNamePrefixFeaturesOutMixin` to NCA
- Added `self._n_features_out` to NCA

Micky774 committed:
Added full calculation of `_n_features_out` in NCA

Micky774 commented on sklearn/neighbors/_base.py:
```diff
@@ -322,7 +322,9 @@ def _radius_neighbors_from_graph(graph, radius, return_distance):
 
 
-class NeighborsBase(MultiOutputMixin, BaseEstimator, metaclass=ABCMeta):
+class NeighborsBase(
```
Moved mixin to specific transformers, and added to `NeighborhoodComponentsAnalysis` as well. They pass the tests now. Thank you for your feedback!

Micky774 commented:
> For `KNeighborsTransformer`, `n_neighbors` is fixed a priori. That should answer your comment above.

I guess my question is more-so: since the outputs of `KNeighborsTransformer` and `RadiusNeighborsTransformer` are distance graphs, is it sufficient to treat them as returning `n_samples_fit_` output features, and to then assign them class-prefixed feature names?

Micky774 committed:
Updated whats_new doc

thomasjpfan commented:
> I guess my question is more-so: since the outputs of KNeighborsTransformer and RadiusNeighborsTransformer are distance graphs, is it sufficient to treat them as returning n_samples_fit_ output features, and to then assign them class-prefixed feature names, as does my current implementation?

Yup, the number of feature names out must match the amount of output features from `transform`.

thomasjpfan commented on sklearn/neighbors/_nca.py:
```diff
@@ -195,6 +197,7 @@ def __init__(
         self.callback = callback
         self.verbose = verbose
         self.random_state = random_state
+        self._n_features_out = n_components
```
We can not set private attributes in `__init__`. We have common checks that test for this behavior. You can see if on the [CI failing tests](https://dev.azure.com/scikit-learn/scikit-learn/_build/results?buildId=36979&view=logs&j=dde5042c-7464-5d47-9507-31bdd2ee0a3a&t=4bd2dad8-62b3-5bf9-08a5-a9880c530c94).

Micky774 replied:
Now passes the tests (at least on my machine)

thomasjpfan commented on sklearn/neighbors/_base.py:
```diff
@@ -469,20 +470,23 @@ def _fit(self, X, y=None):
             self._tree = X._tree
             self._fit_method = X._fit_method
             self.n_samples_fit_ = X.n_samples_fit_
+            self._n_features_out = self.n_samples_fit_
```
`_n_features_out` is not going to be used by non-transformers. We should only set it in the transformers. For example:

https://github.com/scikit-learn/scikit-learn/blob/9daed7f8970c25afaa54f671815afd75605acf26/sklearn/neighbors/_graph.py#L376

Micky774 replied:
Done!

thomasjpfan replied:
Did you push the changes? it looks like `_n_features_out` is still defined in `NeighborsBase`

ogrisel replied:
This line should still be suppressed.

ogrisel replied:
Same comment for the other similar lines below in `NeighborsBase._fit`. They should be done only in `fit` / `_fit` methods of the subclasses that actually need it.

Micky774 replied:
Change is pushed! Had it stashed for some reason -- thanks for catching that!

Micky774 committed:
Moved `_n_features_out` creation to `fit` methods

ogrisel reviewed (commented):
https://github.com/scikit-learn/scikit-learn/pull/22212/files#r785147042 still needs to be addressed. Once done, I think this PR LGTM.

Micky774 committed:
Merge branch 'main' into feature_names_neighbors

Micky774 committed:
Moved `self._n_features_out` calculation to transformers in `neighbors`

thomasjpfan commented on sklearn/neighbors/_nca.py:
```diff
@@ -216,6 +218,7 @@ def fit(self, X, y):
         # Verify inputs X and y and NCA parameters, and transform a copy if
         # needed
         X, y, init = self._validate_params(X, y)
+        self._n_features_out = self.n_components or X.shape[1]
```
I think it's more maintainable to get the feature names out directly from `components_`:

https://github.com/scikit-learn/scikit-learn/blob/1d697842a36fbd93e89d7756bf899f527baf6323/sklearn/neighbors/_nca.py#L251

where `_n_features_out = components_.shape[0]`, because `components_` is used directly in `transform` to compute the output. This way there is less mental overhead to go from `components_` -> `_n_features_out`.

thomasjpfan commented on sklearn/neighbors/_base.py:
```diff
@@ -521,7 +521,6 @@ def _fit(self, X, y=None):
         self._fit_method = self.algorithm
         self._fit_X = X
         self.n_samples_fit_ = X.shape[0]
-
```
This line removal is not related. May you revert this?

(It makes it easier to run `git blame`.)

thomasjpfan reviewed (commented):
Thanks for the update @Micky774 !

We need tests to check the actual names of `get_feature_names_out`. See:

https://github.com/scikit-learn/scikit-learn/blob/1d697842a36fbd93e89d7756bf899f527baf6323/sklearn/cross_decomposition/tests/test_pls.py#L613-L614

Micky774 committed:
Improved `nca._n_features_out` strategy

Micky774 committed:
Fixed typo in calculating `NCA._n_features_out`

Micky774 committed:
Added unit tests for `get_feature_names_out` in `neighbors` module

Micky774 committed:
Updated test names

Micky774 commented:
Added in the unit tests! Thank you for all your active feedback, and for your patience in this. Let me know if there's anything else I should add here!

ogrisel committed:
Merge branch 'main' into feature_names_neighbors

ogrisel reviewed (commented):
I merged the `main` branch to resolve the conflict in the changelog. I also made this an enhancement instead of an API change because those features names method did not exist under a different name in the last released scikit-learn version.

One more code style detail, but otherwise, LGTM.

ogrisel reviewed (approved):
I wanted to approve this PR in previous review, instead of just commenting.

Micky774 committed:
Update sklearn/neighbors/tests/test_nca.py

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

thomasjpfan reviewed (approved):
LGTM
